### ISSUE TITLE
Tutorial deep_atlas shows a huge negative loss in PyTorch 22.09-23.03 on some GPUs

### ISSUE THREAD

**Issue by mingxin-zheng** [Project-MONAI/tutorials#1336]
**Describe the bug**

On V100/A100, the tutorial deep_atlas is reporting correct number on the reg_loss in the 42nd cell when printing the losses
```
print(f"\n\nBest reg_net validation loss: {best_reg_validation_loss}")
print(f"Best seg_net validation loss: {best_seg_validation_loss}")
```

![image](https://user-images.githubusercontent.com/18563433/233930541-ea027cb8-0957-429b-a491-85f965de7e29.png)

On Geforce 3080Ti/3090Ti and L40, the reg loss is a large negative number, for example:

![image](https://user-images.githubusercontent.com/18563433/233930864-966e0498-8213-4d14-88d5-99d0b152650b.png)

**To Reproduce**
1. Install MONAI 1.0.1 or 1.2rc4 on Nvidia PyTorch 22.09 or 23.03 container

**Comment by binliunls** [comment]
Hi @wyli ,
    When I ran this tutorial on the 3090Ti, I got some negative values in tensor `t2_sum - t_avg*t_sum` in this [line](https://github.com/Project-MONAI/MONAI/blob/d421efcedbe04aab4adf293ed1a365eafa99edad/monai/losses/image_dissimilarity.py#L150) as shown below. I think this should not be correct and the values should always be positive. What do you think about this?

![Screenshot 2023-05-12 at 19 16 08](https://github.com/Project-MONAI/tutorials/assets/107988372/dd32facc-3adb-42f1-923b-a62055a384c8)

Thanks,
Bin

**Comment by wyli** [comment]
agreed, looks like some precision issue, I think this should be reported back to the framework team if we can narrow it down to a minimal reproducible example... perhaps you can try to save the inputs to the lncc_loss with `torch.save`

**Comment by binliunls** [comment]
I think it is a NGC pytorch docker related precision problem. Here is the code to show the bug.

```
import torch
import torch.nn.functional as F

x = (1000 * torch.rand((2, 1, 96, 96, 96)) + 2000).to("cuda")
kernel = torch.tensor([[[[[1.0]], [[1.0]], [[1.0]]]]]).to(x)

print(x[1, 0, 68 : 68 + 3, 57 : 57 + 3, 43 : 43 + 3])
torch_y = F.conv3d(x, weight=kernel, groups=1)
print(torch_y[1, 0, 68 : 68 + 3, 57 : 57 + 3, 43 : 43 + 3])
```

Here are three experiments.
1. On a **3090Ti with NGC pytorch 23.03 container**, the output is:

```
tensor([[[2029.1384, 2570.6914, 2323.4832],
         [2064.0557, 2005.5037, 2162.0898],
         [2641.8159, 2601.0364, 2225.3689]],

        [[2876.4146, 2757.2285, 2815.6948],
         [2466.1958, 2672.7217, 2448.4229],
         [2504.8730, 2058.9844, 2685.1572]],

        [[2727.1165, 2921.5139, 2614.0007],
         [2947.5464, 2541.1526, 2222.2314],
         [2725.5344, 2711.1599, 2318.3848]]], device='cuda:0')
tensor([[[7633., 8250., 7754.],
         [7478., 7220., 6832.],
         [7872., 7372., 7230.]],

        [[7906., 7882., 8080.],
         [7496., 8058., 6794.],
         [7520., 7726., 7150.]],

        [[7882., 8092., 7326.],
         [7812., 8138., 7052.],
         [7106., 7838., 7352.]]], device='cuda:0')
```

2. On an **A100 with NGC pytorch 23.03 container**, the output is:

```
tensor([[[2517.1548, 2580.9873, 2951.2512],
         [2364.4739, 2738.0535, 2832.5974],
         [2747.3569, 2441.2136, 2390.8936]],

        [[2133.9880, 2310.7104, 2135.4763],
         [2145.6643, 2203.1594, 2476.5459],
         [2154.6860, 2935.8662, 2883.0042]],

        [[2978.5386, 2900.1516, 2377.1538],
         [2134.7727, 2126.7642, 2282.2429],
         [2583.7375, 2502.0923, 2025.9812]]], device='cuda:0')
tensor([[[7629.6812, 7791.8496, 7463.8813],
         [6644.9111, 7067.9771, 7591.3867],
         [7485.7803, 7879.1724, 7299.8789]],

        [[7979.4766, 8132.9263, 6669.7715],
         [6332.3086, 6539.3042, 6886.6982],
         [6846.7588, 7558.3018, 7401.6973]],

        [[8219.2246, 7927.0127, 6784.8770],
         [6370.7451, 6822.0571, 6976.0708],
         [7141.0293, 6736.4033, 7233.6084]]], device='cuda:0')
```

3.  On a **3090Ti with pytorch official docker** [2.0.1-cuda11.7-cudnn8-devel](https://hub.docker.com/layers/pytorch/pytorch/2.0.1-cuda11.7-cudnn8-devel/images/sha256-4f66166dd757752a6a6a9284686b4078e92337cd9d12d2e14d2d46274dfa9048?context=explore), the output is:

```
tensor([[[2240.3303, 2075.5498, 2538.9990],
         [2020.3495, 2949.1833, 2562.8066],
         [2953.9673, 2862.9373, 2306.1416]],

        [[2983.8008, 2721.1699, 2985.6560],
         [2881.5356, 2585.0393, 2431.8257],
         [2696.4932, 2884.8462, 2565.8350]],

        [[2607.5532, 2881.2222, 2509.1663],
         [2380.2144, 2938.5671, 2737.8425],
         [2842.1265, 2368.8350, 2586.2090]]], device='cuda:0')
tensor([[[7831.6841, 7677.9419, 8033.8213],
         [7282.0996, 8472.7900, 7732.4746],
         [8492.5869, 8116.6182, 7458.1855]],

        [[8102.2036, 8414.7842, 7998.8975],
         [7932.3608, 7631.5518, 7786.7246],
         [8419.9707, 7440.6006, 7655.5684]],

        [[7457.9512, 8087.4541, 7937.7266],
         [7315.0195, 7586.4971, 7983.8232],
         [7940.5142, 7194.7715, 7260.3657]]], device='cuda:0')
```

It seems that the NGC pytorch 23.03 container on the 3090Ti just rounds the output number somehow.  And this leads to a large round-off error, since this logic is recursively used in the [separable_filtering](https://github.com/Project-MONAI/MONAI/blob/e084885c6d498ab02df8107d23aa159f742a2cdd/monai/losses/image_dissimilarity.py#L129) function of the loss function.

**Comment by wyli** [comment]
thanks, maybe you should fix the random seed or try to use exactly the same input to make it easier to see the differences? and perhaps set the datatypes and  the print options to make sure they are not just printing differences https://pytorch.org/docs/stable/generated/torch.set_printoptions.html

**Comment by binliunls** [comment]
> thanks, maybe you should fix the random seed or try to use exactly the same input to make it easier to see the differences? and perhaps set the datatypes and the print options to make sure they are not just printing differences https://pytorch.org/docs/stable/generated/torch.set_printoptions.html

Hi @wyli ,
    Thank you for the advices! I have updated the comment accordingly.

Thanks,
Bin

**Comment by binliunls** [comment]
One more thing. I think this relates to the `tf32` precision of the cuda computation. If I set the `tf32` precision off in the first experiment mentioned above, the output becomes normal and the tutorial could run successfully.

These two lines need to be added to the program to turn off the `tf32` precision.
```python
torch.backends.cuda.matmul.allow_tf32 = False
torch.backends.cudnn.allow_tf32 = False
```

**Comment by wyli** [comment]
so do you think setting `allow_tf32=False` will address the original issue?

**Comment by binliunls** [comment]
Sorry, I forgot to add the tutorial result. I ran the deep atlas tutorial on the 3090Ti NGC pytorch docker with `allow_tf32=False`. The output turned out normally to me. Haven't got a chance to run it on L40 and 3080Ti.

**Comment by wyli** [comment]
ok thanks, we can add `allow_tf32=False` to the notebook as a solution for this ticket, what do you think?

**Comment by wyli** [comment]
created a ticket to track the issue https://github.com/Project-MONAI/MONAI/issues/6525, feel free to create a PR if you have a solution for it...
